### DISCUSSION TITLE
foundry cast call command not working as it should.

### DISCUSSION THREAD

**Discussion by Shourya4641** [Cyfrin/foundry-full-course-cu#2124] [Q&A]
The foundry cast call command is not returning the hexadecimal value for the retrieve function in the SimpleStorage Smart Contract. 
On the one hand, the forge send command for setting the favorite number is getting executed successfully but on the other hand, the call command is not returning the hexadecimal value of the favorite number.

![Screenshot from 2024-07-27 21-55-22](https://github.com/user-attachments/assets/5ba24966-a16a-433f-bae2-4f3024242a4e)
![Screenshot from 2024-07-27 21-55-48](https://github.com/user-attachments/assets/01750acb-26cd-4d2c-b21b-9288d4dc8273)

The screenshots of the terminal are attached.

**Comment by EngrPips** [comment]
Hello @Shourya4641, This has been an issue with `Foundry` since an update some time ago, but I am unsure if they fixed it already. Please look at these two discussions to see if you will see a suggestion to help you solve the issue. 

https://github.com/Cyfrin/foundry-full-course-cu/discussions/1771#discussion-6757027       https://github.com/Cyfrin/foundry-full-course-cu/discussions/459

**Reply by Shourya4641** [reply]
Thank you @EngrPips, I got the output in the cast call command.
I used the broadcast flag while deploying the contract on the test network and then I passed the value. 
After that when I used the cast call command, it returned the hexadecimal value.

**Reply by EngrPips** [reply]
Awesome, Keep crushing it, buddy.

**Reply by 0xCK03** [reply]
Hi @Shourya4641, I'm still facing the same issue even i deployed the contract using the broadcast flag.. Do you have any additional insights please?

**Reply by Shourya4641** [reply]
Hi @0xCK03, make sure to setup the .env file properly using `source .env`. 
After that run the command - 
`cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "store(uint256)" 4641 --rpc-url $RPC_URL --private-key $PRIVATE_KEY --broadcast`.
Then try to run the **call** command, it should run, because I did the same and it worked for me.

**Reply by EngrPips** [reply]
Hey @0xCK03, let us know if @Shourya4641  suggestion works for you.

**Reply by 0xCK03** [reply]
Hi @EngrPips @Shourya4641 , I just done testing and the only way this worked for me is that i deployed the contract using `forge create` not with the deployment script.

**Reply by EngrPips** [reply]
Interesting, @Shourya4641  it worked for you deploying the contract with your deployment script, right?

**Reply by Shourya4641** [reply]
Yes, it did. @EngrPips

**Reply by Shourya4641** [reply]
@0xCK03 , what error are you getting exactly?!

**Reply by 0xCK03** [reply]
@Shourya4641 I'm not getting any errors, It's just output 0x when i use the address of the contract deployed by the script.

**Reply by EngrPips** [reply]
it's weird it doesn't work for you with deployment script but it works for @Shourya4641

**Reply by aliabbas827** [reply]
Hi @Shourya4641 @EngrPips @0xCK03. I think I have found a solution. 

### WITH FORGE CREATE COMMAND:

When I deploy my contract using `forge create` command, it gives me a `Deployed to` address

![image](https://github.com/user-attachments/assets/7c48ca8c-24c0-4f37-b699-422afdbfe347)

Then I use `cast send` command using this `Deployed to` address and it completes the transaction and after that I use `cast call` command with same `Deployed to` address to call `retrieve` and it gives me the hexadecimal number

![image](https://github.com/user-attachments/assets/a2554772-42cb-4414-9391-914984cb0759)


### WITH FORGE SCRIPT COMMAND:

when I use `forge script` command, it give a `contractAddress` in run-latest.json file:

![image](https://github.com/user-attachments/assets/b9869c9b-7371-428f-aa6f-2e2f20f784b2)

Then I use this `contractAddress` in `cast send` command and it completes the transaction and after that I use `cast call` command with same  `contractAddress` to call `retrieve` and it gives me the hexadecimal number

![image](https://github.com/user-attachments/assets/717cc2e6-9474-4575-929f-754af24ef06c)

I hope this helps. 
PS: I used cast wallet to encrypt my private key.

**Reply by EngrPips** [reply]
Interesting. Thanks for this @aliabbas827. Does @aliabbas827's suggestion help you @0xCK03?
